packages/integration-sdk-http-client/src/index.ts

```diff
@@ -244,10 +244,84 @@ export abstract class BaseAPIClient {
+   * This method is designed to facilitate paginated API requests where each request's response determines the parameters of the next request.
+   *
+   * @typeparam T - The expected type of the response object used in the iteration callback.
+   *
+   * @param cb - A callback function that is called after each API request. It receives the API response (`iterationResources`) of type `T` and returns an object containing `nextRequestOptions` and `nextRequestQuery` for the next iteration, along with a `hasNext` boolean indicating whether to continue iterating.
+   *   @param iterationResources - The response object from the current API request.
+   *   @returns An object with:
+   *     - `nextRequestOptions`: Optional RequestOptions for the next API request.
+   *     - `nextRequestQuery`: An optional object representing the query parameters for the next API request.
+   *     - `hasNext`: A boolean indicating whether there are more requests to be made.
+   *
+   * @param initialRequest - An object containing the initial API request details.
+   *   @param endpoint - The endpoint URL of the initial API request.
+   *   @param options - Optional RequestOptions for the initial API request.
+   *
+   * @returns A promise that resolves when all iterations are complete. The promise does not resolve with a specific value, indicating the primary purpose of this method is to perform side effects (e.g., accumulating data, side effects in `cb`, etc.).
+   *
+   * @example
+   * // Assuming an API where we fetch paginated data
+   * const fetchPaginatedData = async () => {
+   *   await iterateApi<MyResponseType>(async (response) => {
+   *     const hasNextPage = response.page < response.totalPages;
+   *     const nextQuery = hasNextPage ? { page: response.page + 1 } : undefined;
+   *
+   *     return {
+   *       hasNext: hasNextPage,
+   *       nextRequestQuery: nextQuery,
+   *     };
+   *   }, {
+   *     endpoint: 'https://api.example.com/data',
+   *     options: { method: 'GET' },
+   *   });
+   * };
+   *
+   */
+
+  protected async iterateApi<T>(
+    cb: (iterationResources: T) => {
+      nextRequestOptions?: RequestOptions;
+      nextRequestQuery?: { [key: string]: string };
+      hasNext: boolean;
+    },
+    initialRequest: {
+      endpoint: string;
+      options?: RequestOptions;
+    }
+  ): Promise<void> {
+    let hasNext = false;
+    let nextRequestQuery: { [key: string]: string } | undefined;
+    let nextRequestOptions: RequestOptions | undefined;
+
+    do {
+      const response = await this.request(
+        nextRequestQuery
+          ? `${initialRequest.endpoint}${this.buildQueryParams(
+              nextRequestQuery
+            )}`
+          : initialRequest.endpoint,
+        nextRequestOptions ? nextRequestOptions : initialRequest.options
+      );
+
+      const cbOptions = cb(response as T);
```

Comment:
You don't need a generic in the function at all if you are returning the `Response` object (which we should, so the package clients have more control), so instead of `cb: (iterationResources: T) => {` it should be `cb: (response: Response) => {`
Unfortunately the Response type doesn't accept generics